docs/gnmi_extensions.md

```diff
@@ -0,0 +1,76 @@
+with the extra functionality to do network wide management of the configuration
+across many devices requires some extension.
+
+Some functions are far beyond what gNMI is designed for, and have been given
+their own gRPC definitons as [admin](../pkg/northbound/proto/admin.proto) and
+[diags](../pkg/northbound/proto/diags.proto).
+
+## Use of target in the NBI
+In gNMI the Path type is comprised of a set of path elements and a **target**.
+In the onos-config NBI this target represents the **name of the device** as it
+is held in the configuration system. Configurations can be explored through the
+```bash
+onos configs
+```
+
+The following rules apply when a **prefix** is present in the request:
+* The target in the prefix always takes precedence over any others
+* If no target is given in the prefix it is an error
+
+### Special case to get all device names
+When doing a GetRequest if '*' is given as a **target** then the request returns a simple listing of all device names present in the system (with their version). Any path elements are ignored in this special case.
+
+
+# Managing configuration objects
+The diagram shows the internal storage structures of onos-config (in orange). The
+Configuration object represents the complete configuration for a particular
+version of a device.
+
+Network changes can be applied across multiple Configurations (devices), and have
+the ability to be rolled back (by the name of the Network Change).
+ 
+![onos-config internals](images/onos-config-internals.png)
+
+3 different extensions have been chosen in the project to make dealing with Network
+Changes and Configurations through gNMI possible.
+
+## Use of Extension 100 (network change name) in SetRequest and SetResponse
+In onos-config the gNMI extension number 100 has been reserved for the
+**network change name**.
+
+### SetRequest
+In the SetRequest extension 100 can be used to define a name for the Network
+change. If it is not specified then a name is picked automatically.
+> There is an example of setting this extension when using gnmi_cli in
+[gnmi.md](gnmi.md) (Northbound Set Request via gNMI)
+
+### SetResponse
+In the SetResponse the name of the Network Change will always be given in
+extension 100 (either the given name or the generated one).
+>There is an example of the return of this extension through **gnmi_cli** in
+[gnmi.md](gnmi.md) (Northbound Set Request via gNMI)
+
+## Use of Extension 101 (device version) in SetRequest
+Extension 101 is used to set the Model version for a Configuration (as part of a
+Network Change). There may be multiple different configurations for a device based
+on version number. This extension allows the correct version of the configuration
+to be chosen.
+
+If a version is given where no existing Configuration exists, an error is
+returned explaining that device type (extension 102 - see below) should also be
+specified.
+
+If no extension 101 (version) is given, and only one Configuration already exists
```

Comment:
I believe that if no 101 is given and no config exists one gets created, correct?

Reply:
It can't create a config from new without a version **and** a type so 101 and 102 are both needed. The code might not be up to this yet, but will have to.

Reply:
ok